File: togo/internal/cli/update.go
package cli

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/nikeee/cappu/internal/config"
	"github.com/nikeee/cappu/internal/install"
	"github.com/nikeee/cappu/internal/sources"
)

// RunUpdate handles `cappu update`: bump every declared dependency to the
// newest stable version that keeps its configuration's transitive graph
// conflict-free, rewrite cappu.json (comments preserved), then refresh the lock
// via install. Port of src/cli/update.ts.
func RunUpdate(configPathArg string, cfg *config.Config) int {
	if !cfg.FromFile {
		fmt.Fprintln(os.Stderr, "cappu: no cappu.json found - run `cappu init` first")
		emitAnnotation("error", "no cappu.json found - run `cappu init` first", AnnotationLocation{})
		return 1
	}

	errp := painter(os.Stderr)
	fmt.Fprint(os.Stderr, errp("cyan", "checking for updates...\n"))

	bumps, err := install.PlanUpdates(cfg, sources.Configured(cfg))
	if err != nil {
		fmt.Fprintf(os.Stderr, "cappu: update failed: %s\n", err)
		emitAnnotation("error", fmt.Sprintf("update failed: %s", err), AnnotationLocation{})
		return 2
	}
	if len(bumps) == 0 {
		fmt.Fprintln(os.Stdout, "all dependencies are up to date")
		return 0
	}

	configPath := filepath.Join(cfg.BaseDir, config.DefaultConfigName)
	if configPathArg != "" {
		if abs, e := filepath.Abs(configPathArg); e == nil {
			configPath = abs
		}
	}
	text, err := os.ReadFile(configPath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "cappu: %s\n", err)
		return 1
	}
	for _, b := range bumps {
		text, err = config.SetDependency(text, b.Configuration, b.Key, b.To)
		if err != nil {
			fmt.Fprintf(os.Stderr, "cappu: %s\n", err)
			return 1
		}
	}
	if err := os.WriteFile(configPath, text, 0o644); err != nil {
		fmt.Fprintf(os.Stderr, "cappu: %s\n", err)
		return 1
	}
	for _, b := range bumps {
		fmt.Fprintf(os.Stderr, "updated %s: %s -> %s\n", b.Key, b.From, b.To)
	}

	reloaded, err := config.Load(configPath, "")
	if err != nil {
		fmt.Fprintf(os.Stderr, "cappu: %s\n", err)
		return 1
	}
	return runInstallWith(reloaded, false, true)
}

File: src/cli/update.ts
// `cappu update`: bump every declared dependency to the newest stable version
// that keeps its configuration's transitive graph conflict-free, rewrite
// cappu.json (preserving comments), then refresh the lock via install.

import { readFileSync, writeFileSync } from "node:fs";
import { join, resolve } from "node:path";

import { type CappuConfig, DEFAULT_CONFIG_NAME, loadConfig } from "../config.ts";
import { type DependencyBump, planUpdates } from "../install.ts";
import { type PackageSource } from "../packages/index.ts";
import { emitAnnotation } from "./annotations.ts";
import { runInstall } from "./install.ts";
import { hasJsoncKey, setJsoncValue } from "./jsoncEdit.ts";
import { painter } from "./style.ts";

/** Overwrite the bumped versions in the JSONC config text, comments intact. */
export function applyBumpsToJsonc(text: string, bumps: readonly DependencyBump[]): string {
  for (const bump of bumps) {
    const path = ["dependencies", bump.configuration, bump.key];
    // Only overwrite an entry that is still declared (a vanished section is
    // skipped, never recreated).
    if (hasJsoncKey(text, path)) text = setJsoncValue(text, path, bump.to);
  }
  return text;
}

export async function runUpdate(
  configPathArg: string | undefined,
  config: CappuConfig,
  // Injectable for tests; defaults to the configured remote repositories.
  sources?: readonly PackageSource[],
): Promise<never> {
  if (!config.fromFile) {
    process.stderr.write("cappu: no cappu.json found - run `cappu init` first\n");
    emitAnnotation("error", "no cappu.json found - run `cappu init` first");
    process.exit(1);
  }

  const err = painter(process.stderr);
  process.stderr.write(err("cyan", "checking for updates...\n"));

  let bumps: DependencyBump[];
  try {
    bumps = await planUpdates(config, sources);
  } catch (e) {
    process.stderr.write(`cappu: update failed: ${(e as Error).message}\n`);
    emitAnnotation("error", `update failed: ${(e as Error).message}`);
    process.exit(2);
  }

  if (bumps.length === 0) {
    process.stdout.write("all dependencies are up to date\n");
    process.exit(0);
  }

  const configPath = configPathArg
    ? resolve(configPathArg)
    : join(config.baseDir, DEFAULT_CONFIG_NAME);
  writeFileSync(configPath, applyBumpsToJsonc(readFileSync(configPath, "utf8"), bumps));
  for (const b of bumps) {
    process.stderr.write(`updated ${b.key}: ${b.from} -> ${b.to}\n`);
  }

  // Re-resolve and rewrite the lock against the bumped versions.
  return runInstall(loadConfig(configPath), { updateLock: true });
}
